atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java

```diff
@@ -0,0 +1,181 @@
+        SortedMap<Token, TokenRange> tokenRangesByEnd =
+                KeyedStream.of(tokenRanges).mapKeys(TokenRange::getEnd).collectTo(TreeMap::new);
+        Set<TokenRange> ranges = getMinimalSetOfRangesForTokens(metadata, partitionKeyTokens, tokenRangesByEnd);
+        Multimap<Host, TokenRange> tokenMapping = ArrayListMultimap.create();
+        ranges.forEach(range -> {
+            List<Host> hosts = ImmutableList.copyOf(metadata.getReplicas(quotedKeyspace(keyspace), range));
+            if (hosts.isEmpty()) {
+                throw new SafeIllegalStateException(
+                        "Failed to find any replicas of token range for repair",
+                        SafeArg.of("tokenRange", range.toString()),
+                        SafeArg.of("keyspace", quotedKeyspace(keyspace)));
+            }
+            hosts.forEach(host -> tokenMapping.put(host, range));
+        });
+        return KeyedStream.stream(tokenMapping.asMap())
+                .map(Collection::stream)
+                .map(stream -> stream.collect(Collectors.toList()))
+                .collectToMap();
+    }
+
+    // VisibleForTesting
+    public static Set<TokenRange> getMinimalSetOfRangesForTokens(
+            Metadata metadata, Set<Token> partitionKeyTokens, SortedMap<Token, TokenRange> tokenRangesByEnd) {
+        Map<Token, TokenRange> tokenRangesByStartToken = new HashMap<>();
+        for (Token token : partitionKeyTokens) {
+            TokenRange minimalTokenRange = findTokenRange(metadata, token, tokenRangesByEnd);
+            tokenRangesByStartToken.merge(
+                    minimalTokenRange.getStart(), minimalTokenRange, ClusterMetadataUtils::findLatestEndingRange);
+        }
+        return ImmutableSet.copyOf(tokenRangesByStartToken.values());
+    }
+
+    private static TokenRange findTokenRange(
+            Metadata metadata, Token token, SortedMap<Token, TokenRange> tokenRangesByEnd) {
+        if (tokenRangesByEnd.containsKey(token)) {
+            return tokenRangesByEnd.get(token);
+        } else if (!tokenRangesByEnd.headMap(token).isEmpty()) {
+            return metadata.newTokenRange(tokenRangesByEnd.headMap(token).lastKey(), token);
+        } else {
+            // Confirm that the first entry in the sorted map is the wraparound range
+            TokenRange firstTokenRange = tokenRangesByEnd.get(tokenRangesByEnd.firstKey());
+            Preconditions.checkState(
+                    firstTokenRange.isWrappedAround(),
+                    "Failed to identify wraparound token range",
+                    SafeArg.of("firstTokenRange", firstTokenRange),
+                    SafeArg.of("token", token));
+            return metadata.newTokenRange(firstTokenRange.getStart(), token);
+        }
+    }
+
+    private static String quotedKeyspace(String keyspaceName) {
+        return "\"" + keyspaceName + "\"";
+    }
+
+    // Expects all ranges to have the same start token
```

Comment:
This can also be a precondition check instead of a comment

Reply:
it already is a precondition check :)

Reply:
removed the redundant comment